greetings: regularly print a string to the console every second.

Flameeyes opened the pull request:
Introduce a new sample app to repeatedly and regularly print a string
("Zephyr says 'Hi!'") every second on the console.

Recommend this sample instead of hello_world for boards that are not
meeting the minimum requirements for blinky.

kartben reviewed (changes requested):
Good idea. Could you please also update the README ("Overview" and "Sample Output")? Thanks!

gmarull commented on samples/hello_world/src/main.c:
```diff
@@ -6,8 +6,14 @@
-	printk("Hello World! %s\n", CONFIG_BOARD);
+	while(1) {
+		printk("Hello World! %s\n", CONFIG_BOARD);
+		k_msleep(SLEEPTIME);
```
```suggestion
		k_msleep(1000);
```

gmarull replied:
just for simplicity, also, I'd change `printk` to `printf`, and include `<stdio.h>`

gmarull commented on samples/hello_world/src/main.c:
```diff
@@ -6,8 +6,14 @@
 
 #include <zephyr/kernel.h>
 
+/* delay between greetings (in ms) */
+#define SLEEPTIME 1000
+
```
```suggestion
```

nashif reviewed (dismissed):
just use `samples/synchronization`/ sample. Lets not change the behavior of a classic hello world sample.

Flameeyes commented:
> just use `samples/synchronization`/ sample. Lets not change the behavior of a classic hello world sample.

I did that originally, but that does increase the noise surface significantly, since it's involving threading (attach debugger, where are you?)

During bring up of a new board (particularly if, like me, you have no idea what you're doing and is the first time you try), being able to exclude as much of the higher level stack is useful. Already bringing up the serial console adds more moving parts, which is why at least for the local testing I preferred reducing the surface area.

Would a trade-off of having this `#define` controllable be worth exploring?

fabiobaltieri commented:
I also think it's a good idea to have it repeat, then you'd still get an output if you are fiddling with serial ports and end up missing the boot message for whatever reason.

Flameeyes committed:
greetings: regularly print a string to the console every second.

Introduce a new sample app to repeatedly and regularly print a string
("Zephyr says 'Hi!'") every second on the console.

Recommend this sample instead of hello_world for boards that are not
meeting the minimum requirements for blinky.

Signed-off-by: Diego Elio Pettenò <flameeyes@meta.com>

nashif commented:
> I also think it's a good idea to have it repeat, then you'd still get an output if you are fiddling with serial ports and end up missing the boot message for whatever reason.

again, we have the syncronization sample for this :)

kartben commented:
> > I also think it's a good idea to have it repeat, then you'd still get an output if you are fiddling with serial ports and end up missing the boot message for whatever reason.
> 
> again, we have the syncronization sample for this :)

TBH it is a bit of a stretch to compare improving the Hello World sample by adding a simple sleep/while(1) to a full blown synchronization sample that demonstrates how to create threads, use semaphores, etc. Should we refrain from using k_sleep in every sample then, for this should be something reserved only for the synchronization sample? :)

I think the point here is really about improving the usability/UX of the Hello World (in addition to also making it a quick proof point that clock is setup properly, to @Flameeyes' point earlier), and nothing more.

fabiobaltieri reviewed (dismissed):
I'm with Ben on this, `synchronization` is good for showing thread usage but that's few steps ahead and this PR made me realize how may times I've hacked a printk into `blinky` while searching for the TX pin on whatever board I had just for this.

Though it is indeed sad to diverge from the classic version of it, we were already a bit off from the start, it should have been a `printf` and all lowercase.

![x](https://github.com/zephyrproject-rtos/zephyr/assets/891546/51d1fd40-d63f-4e90-87b5-6b23f4ca034f)

nashif commented:
https://www.geeksforgeeks.org/hello-world-in-30-different-languages/

hello world is a defact "standard", it has been the same code in Zephyr since we started, lets not change that just because of someone's debug needs, if you want to debug something, take hello world and modify it.

Flameeyes commented:
Let me try to find a way forward:

 * there seem to be a want, besides from me, for a simple, regular output on the UART for various debugging purposes;
 * what made sense to me, is the equivalent of blinky, that is to output something every second or so — I have done that on hello_world because that's what [the documentation](https://docs.zephyrproject.org/latest/develop/getting_started/index.html#build-the-blinky-sample) suggests using it instead of blinky if the latter can't be used;
 * I personally believe that synchronization is not an equivalent simple app to blinky due to the aforementioned additional surface area
 * there's precedent for hello_world to simply terminate and not do anything more

So here's an alternative approach I'm happy to run with if that raise no complaints: introduce a new application that is, semantically "blinky but on a serial port", update the documentation to suggest the use of that instead of hello_world where blinky can't be uesd.

Maybe something with a repeating pattern like the old C64 demos? Like "\/" sent very one second? There's a bit more use in slightly longer patterns as they're easier to caught with the debugger though, so maybe a `"Zephyr says 'Hi!'"` line every second?

nashif commented:
> So here's an alternative approach I'm happy to run with if that raise no complaints: introduce a new application that is, semantically "blinky but on a serial port", update the documentation to suggest the use of that instead of hello_world where blinky can't be uesd.

that works, instead of changing the existing hello_world, you can add a new one implementing similar functionality to blinky for development/debugging needs.

gmarull commented:
so what about blinky printing every second as well?

Flameeyes commented:
> so what about blinky printing every second as well?

blinky has a minimum hardware requirement for the led0 alias, or it'll fail to build — which is where the recommendation for hello_world comes from. 

I mean, we could still have it print as well to combine that option, but complicating the sample to make led0 optional doesn't seem like a good idea.

Flameeyes commented:
Not entirely sure on what's the best practice between re-using the same pull request or separating it, but I think keeping the discussion in one place is worthwhile — happy to submit as a separate pull request if you think it's better.

Flameeyes commented:
More to the need of not relying on philosophers: as of today philosophers causes a hard fault on the EV11L78A because _something_ is broken in the threading code.

@nashif what's the blocker for having this new sample app?

carlescufi commented:
@kartben can you take another look?

fabiobaltieri commented:
@kartben ping

kartben commented on samples/basic/greetings/README.rst:
```diff
@@ -0,0 +1,39 @@
+.. _greetings:
+
+Greetings
+#########
```
```suggestion
.. zephyr:code-sample:: greetings
   :name: Greetings
   :relevant-api: thread_apis

   Print greetings to the console once per second.
```

kartben commented on samples/basic/greetings/README.rst:
```diff
@@ -0,0 +1,39 @@
+A simple sample that can be used with any :ref:`supported board <boards>` and
+prints "Zephyr says 'Hi!'" to the console, once per second.
+
+It's an alternative to :ref:`blinky <blinky-sample>` for boards with no
```
```suggestion
It's an alternative to :zephyr:code-sample:`blinky` for boards with no
```

kartben commented on samples/basic/greetings/README.rst:
```diff
@@ -0,0 +1,39 @@
+This application can be built and executed on QEMU as follows:
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/greetings
```
```suggestion
   :zephyr-app: samples/basic/greetings
```

kartben commented on doc/develop/getting_started/index.rst:
```diff
@@ -652,7 +652,7 @@ Build the Blinky Sample
    Blinky is compatible with most, but not all, :ref:`boards`. If your board
    does not meet Blinky's :ref:`blinky-sample-requirements`, then
-   :ref:`hello_world` is a good alternative.
+   :ref:`greetings` is a good alternative.
```
```suggestion
   :zephyr:code-sample:`greetings` is a good alternative.
```

kartben reviewed (changes requested):
an actual mistake re: sample path + some requested changes to leverage the recently introduced Zephyr sample Sphinx extension. Thanks @Flameeyes!

kartben commented:
@Flameeyes will you have cycles to address the opened comments? Thanks!

Flameeyes commented:
Probably next week, let me see!

kartben commented:
@Flameeyes thanks! FWIW blinky is now *also* printing to the console, so this might actually not be all that necessary to add this code sample? (PR #63056)

Flameeyes commented:
Given this started while trying to get blinky to print, if we settled on that I don't need that.
